### PR TITLE
EPP-4 set up of namespaces and envs

### PR DESCRIPTION
**Pull request by Jholderdevops**
## What? 
https://collaboration.homeoffice.gov.uk/jira/browse/EPP-4 
Creation of a new service - envs and secrets and namespace and 
Adding files and folders necessary to deploy into environments via CI/CD pipeline.

## Why? 
This is the first initial setup of Kubernetes with its secrets the service can be deployed into the namespaces.

## How? 
Adding the right configs and kube files

## Testing?
WIP

## Check list

- [x] I have reviewed my own pull request for linting issues (e.g. adding new lines)
- [ ] I have written tests (if relevant)
- [x] I have created a JIRA number for my branch
- [x] I have created a JIRA number for my commit
- [x] I have followed the chris beams method for my commit https://cbea.ms/git-commit/
here is an [example commit](https://github.com/UKHomeOfficeForms/hof/commit/810959f391187c7c4af6db262bcd143b50093a6e)
- [x] Ensure drone builds are green especially tests
- [x] I will squash the commits before merging

### CODE REVIEW

File: .drone.yml [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+  UAT_ENV: sas-epp-uat
+  BRANCH_ENV: sas-epp-branch
+  PRODUCTION_URL: www.tbc.homeoffice.gov.uk
+  IMAGE_URL: 340268328991.dkr.ecr.eu-west-2.amazonaws.com
+  IMAGE_REPO: sas/epp
+  GIT_REPO: UKHomeOffice/explosives-precursors-poisons
+  HOF_CONFIG: hof-services-config/Explosives_Precursors_Poisons
+  NON_PROD_AVAILABILITY: Mon-Fri 08:00-23:00 Europe/London
+  READY_FOR_TEST_DELAY: 20s
+  NOTIFY_STUB: stub
```
Comment (sulthan-ahmed) [inline .drone.yml:19]: suggest removing things like prod if we don't have it yet and anything else

File: .drone.yml [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+#     ENVIRONMENT: DRONE
+#   volumes:
+#     - name: dockersock
+#       path: /root/.dockersock
+#   commands:
+#     - yarn run test:accessibility
+
+# acceptance_tests: &acceptance_tests
+#   pull: if-not-exists
+#   image: mcr.microsoft.com/playwright:v1.12.3-focal
```
Comment (sulthan-ahmed) [inline .drone.yml:69]: I'm not a big fan of commenting out code

File: .drone.yml [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+    resources:
+      limits:
+        cpu: 1000
+        memory: 1024Mi
+    environment:
+      IMAGE_NAME: sas/epp:${DRONE_COMMIT_SHA}
+      SEVERITY: MEDIUM,HIGH,CRITICAL  --dependency-tree --format table
+      FAIL_ON_DETECTION: false
+      IGNORE_UNFIXED: false
+      ALLOW_CVE_LIST_FILE: hof-services-config/E_Visa_Find_My_Reference_Form/trivy/.trivyignore.yaml
```
Comment (sulthan-ahmed) [inline .drone.yml:212]: remove this, we need one for epp rather than Evisa

File: .drone.yml [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+      branch:
+        include:
+          - master
+          - feature/*
+      event: pull_request
+    depends_on:
+      - clone_repos
+      - image_to_ecr
+
+  - name: deploy_to_uat #TBD
```
Comment (sulthan-ahmed) [inline .drone.yml:243]: Suggest removing the comment.  I appreciate it's a comment to yourself

File: .drone.yml [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+      KUBE_TOKEN:
+        from_secret: kube_token_prod
+    commands:
+      - bin/deploy.sh $${STG_ENV}
+    when:
+      branch: master
+      event: push
+    depends_on:
+      - clone_repos
+      - image_to_ecr
```
Comment (sulthan-ahmed) [inline .drone.yml:316]: suggest remove unless it is part of the work you are doing

File: .drone.yml [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+    image: redis
+    volumes:
+      - name: dockersock
+        path: /var/run
+
+volumes:
+  - name: dockersock
+    temp: {}
+
+...
```
Comment (sulthan-ahmed) [inline .drone.yml:494]: I say keep the PR simple and remove things not part of this PR

File: kube/create_secrets.sh [hunk 7]
```diff
@@ -0,0 +1 @@
+kubectl create secret generic redis --from-literal=session_secret=${SESSION_SECRET}
```
Comment (sulthan-ahmed) [inline kube/create_secrets.sh:1]: Did you mean to add this file?

File: kube/file-vault/file-vault-deployment.yml [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        - name: certs
+          secret:
+            {{ if eq .KUBE_NAMESPACE .BRANCH_ENV }}
+            secretName: branch-tls-external
+            {{ else }}
+            secretName: file-vault-cert
+            {{ end }}
+        - name: bundle
+          configMap:
+            name: bundle
```
Comment (sulthan-ahmed) [inline kube/file-vault/file-vault-deployment.yml:239]: I would not add filevault as part of your PR

**Review by sulthan-ahmed — COMMENTED** [review]
I prefer setting up the infrastructure with minimum code e.g. just the node app and the redis app initially

File: kube/app/deployment.yml [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+          ports:
+            - containerPort: 10443
+          volumeMounts:
+            - mountPath: /public
+              name: public
+          securityContext:
+            runAsNonRoot: true
+      volumes:
+        - name: public
+          emptyDir: {}
```
[the remark this answers is not in the document]
Comment (Jholderdevops) [reply kube/app/deployment.yml:166]: relevant files added to this version of the pipeline

[the remark this answers is not in the document]
Comment (Jholderdevops) [reply kube/app/deployment.yml:166]: will push shortly I think what ever we do not need as part of branch and uat can go

**Comment by sulthan-ahmed** [comment]
You have two commits, can you tick the box about this ` I will squash the commits before merging`

**Comment by sulthan-ahmed** [comment]
Can you rename your commit to match your PR name so that ns is written as namespace? I prefer avoiding acronyms where we can

File: .drone.yml [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+      event: [push, pull_request]
+    depends_on:
+      - build_image
+
+  # Deploy to pull request UAT environment
+  - name: deploy_to_branch
+    pull: if-not-exists
+    image: quay.io/ukhomeofficedigital/kd:v1.18.0
+    environment:
+      # NOTIFY_STUB: stub
```
Comment (sulthan-ahmed) [inline .drone.yml:218]: I suggest removing this comment

File: kube/app/deployment.yml [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+              {{ if eq .KUBE_NAMESPACE .BRANCH_ENV }}
+              value: redis-{{ .DRONE_SOURCE_BRANCH }}
+              {{ else }}
+              value: redis
+              {{ end }}
+#TBD        # - name: NOTIFY_KEY
+            #   valueFrom:
+            #     secretKeyRef:
+            #       name: notify-key
+            #       key: notify-key
```
Comment (sulthan-ahmed) [inline kube/app/deployment.yml:75]: remove comment, I'd say

File: kube/app/deployment.yml [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            # - name: KEYCLOAK_USERNAME
+            #   valueFrom:
+            #     secretKeyRef:
+            #       name: keycloak-user
+            #       key: username
+            # - name: KEYCLOAK_PASSWORD
+            #   valueFrom:
+            #     secretKeyRef:
+            #       name: keycloak-user
+            #       key: password
```
Comment (sulthan-ahmed) [inline kube/app/deployment.yml:97]: remove all comments unless they are legitimately explaining something

File: kube/app/deployment.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (sulthan-ahmed) [reply kube/app/deployment.yml:166]: I can still see some comments, did you mean to keep them?

File: kube/app/ingress-external.yml [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  tls:
+    - hosts:
+      {{ if eq .KUBE_NAMESPACE .BRANCH_ENV }}
+        - {{ .APP_NAME }}-{{ .DRONE_SOURCE_BRANCH }}.branch.sas-notprod.homeoffice.gov.uk
+      {{ else if eq .KUBE_NAMESPACE .UAT_ENV }}
+        - {{ .APP_NAME }}.uat.sas-notprod.homeoffice.gov.uk
+      {{ else if eq .KUBE_NAMESPACE .STG_ENV }}
+        - {{ .APP_NAME }}.stg.sas.homeoffice.gov.uk
+      {{ else if eq .KUBE_NAMESPACE .PROD_ENV }}
+        - {{ .PRODUCTION_URL }}
```
Comment (sulthan-ahmed) [inline kube/app/ingress-external.yml:22]: as discussed I would remove anything about Production

File: kube/app/ingress-internal.yml [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+      secretName: branch-tls-internal
+      {{ else }}
+      secretName: {{ .APP_NAME }}-internal
+      {{ end }}
+  rules:
+    {{ if eq .KUBE_NAMESPACE .BRANCH_ENV }}
+    - host: {{ .APP_NAME }}-{{ .DRONE_SOURCE_BRANCH }}.internal.branch.sas-notprod.homeoffice.gov.uk
+    {{ else if eq .KUBE_NAMESPACE .UAT_ENV }}
+    - host: {{ .APP_NAME }}.internal.uat.sas-notprod.homeoffice.gov.uk
+    {{ else if eq .KUBE_NAMESPACE .STG_ENV }}
```
Comment (sulthan-ahmed) [inline kube/app/ingress-internal.yml:32]: I'd remove staging until we know we are having it

File: kube/file-vault/file-vault-deployment.yml [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+        {{ if eq .KUBE_NAMESPACE .BRANCH_ENV }}
+        name: file-vault-{{ .DRONE_SOURCE_BRANCH }}
+        service: file-vault-{{ .DRONE_SOURCE_BRANCH }}
+        {{ else }}
+        name: file-vault
+        service: file-vault
+        {{ end }}
+    spec:
+      containers:
+        - name: file-vault
```
Comment (sulthan-ahmed) [inline kube/file-vault/file-vault-deployment.yml:38]: I don't believe the first setup should have filevault, it makes it more complicated

File: kube/file-vault/file-vault-deployment.yml (hunk 8 above)
Comment (sulthan-ahmed) [reply kube/file-vault/file-vault-deployment.yml:233]: Did you meant to keep filevault there?

**Review by jamiecarterHO — COMMENTED** [review]
Couple of comments from me

File: kube/app/deployment.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jholderdevops) [reply kube/app/deployment.yml:166]: pushing the newest change - no comments remain

File: kube/create_secrets.sh (hunk 7 above)
Comment (Jholderdevops) [reply kube/create_secrets.sh]: removed

File: kube/file-vault/file-vault-deployment.yml (hunk 8 above)
Comment (Jholderdevops) [reply kube/file-vault/file-vault-deployment.yml:233]: yes I am adding fv i have it configured, we have that as part of the ticket

File: kube/app/deployment.yml (hunk 11 above)
Comment (Jholderdevops) [reply kube/app/deployment.yml:75]: all comments removed

File: kube/file-vault/file-vault-deployment.yml (hunk 15 above)
Comment (Jholderdevops) [reply kube/file-vault/file-vault-deployment.yml:38]: I have it configured, as it is part of the sub-tasks...unless we separate it as a main ticket on its own?

File: Dockerfile [hunk 16]
```diff
@@ -1,11 +1,12 @@
-FROM node:lts-alpine@sha256:19eaf41f3b8c2ac2f609ac8103f9246a6a6d46716cdbe49103fdb116e55ff0cc
+FROM node:20.18.0-alpine3.20@sha256:d504f23acdda979406cf3bdbff0dff7933e5c4ec183dda404ed24286c6125e60
```
Comment (sulthan-ahmed) [inline Dockerfile:1]: this is good I prefer a proper version highlighted so we can keep a track

File: kube/app/deployment.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (sulthan-ahmed) [reply kube/app/deployment.yml:120]: thank you

**Review by sulthan-ahmed — COMMENTED** [review]
Look good, thanks for taking on board my comments.  Just that one minor thing to take out and then happy to approve 🙂

File: bin/deploy.sh [hunk 17]
```diff
@@ -0,0 +1,38 @@
+#! /bin/bash
+set -e
+
+export INGRESS_INTERNAL_ANNOTATIONS=$HOF_CONFIG/ingress-internal-annotations.yaml
+export INGRESS_EXTERNAL_ANNOTATIONS=$HOF_CONFIG/ingress-external-annotations.yaml
+export CONFIGMAP_VALUES=$HOF_CONFIG/configmap-values.yaml
+export NGINX_SETTINGS=$HOF_CONFIG/nginx-settings.yaml
+export FILEVAULT_NGINX_SETTINGS=$HOF_CONFIG/filevault-nginx-settings.yaml
+export FILEVAULT_INGRESS_EXTERNAL_ANNOTATIONS=$HOF_CONFIG/filevault-ingress-external-annotations.yaml
```
[the remark this answers is not in the document]
Comment (Jholderdevops) [reply bin/deploy.sh]: resolved

File: kube/create_secrets.sh (hunk 7 above)
Comment (sulthan-ahmed) [reply kube/create_secrets.sh]: nice one

File: kube/file-vault/file-vault-deployment.yml (hunk 15 above)
Comment (sulthan-ahmed) [reply kube/file-vault/file-vault-deployment.yml]: yeah
you want to do it step by step imho
